### ISSUE TITLE
Unstable diagnostics

### ISSUE THREAD

**Issue by mrkam2** [clj-kondo/clj-kondo#2225]
**version**

`clj-kondo v2023.10.20`

**problem**

When running the clj-kondo multiple times on the same code, it produces slightly different results. For example, when I was running clj-kondo on a large codebase that produced `linting took 57209ms, errors: 688, warnings: 248` results, on the next run (as well as on the several others) it produced `linting took 58501ms, errors: 688, warnings: 247` with only one warning difference: `warning: unused binding app-3`. 

From what I can see, it is a true warning but somehow it is not reliably reported (mostly not reported based on my observations). This `app-3` variable is found in a test that has roughly the following structure:

```clojure
(defdbtest test-name [db]
  (let [...
        app-3 (create-app db ...)
        ...]    
    ; app-3 is not referenced anywhere in the body of the test
    ...))
```

where `defdbtest` is linted as `defn` (it has inline `{:clj-kondo/lint-as 'clojure.core/defn}` definition).

**repro**

I don't know what a minimal reproducible test case is.

**expected behavior**

`unused binding app-3` warning always reported.

What might be the source of such instability? What can I do to diagnose this further?

**Comment by mrkam2** [comment]
I actually run this as a job with the script that runs the following two commands:

```
clj-kondo --lint "$(clojure -Spath)" --copy-configs --dependencies --parallel
clj-kondo --lint <all-src-paths>
```

so it should be quite stable.

How do you think `.clj-kondo/.cache` might be different to produce this exact difference? 

I can do two tests to verify this hypothesis:
1. Remove all `.clj-kondo/.cache` folders if I have any at the beginning of the job.
2. Remove the inline config for the `defdbtest` and see if that triggers the warning.

**Comment by mrkam2** [comment]
Oh, I think I found something that might be the problem. And actually it might be good to safeguard against such cases.

I found that the same macro had two lint-as definitions - one inline and another in the global config.edn file where it was configured with `clj-kondo.lint-as/def-catch-all`.

I guess `--parallel` switch on the `--copy-configs` command caused it to be overridden randomly by one or another config.

Would it be possible to at least issue a warning when configs are in conflict?

It might also be great for clj-kondo to verify the configuration. Given that a macro configuration can be defined in several places, it is easy to misconfigure. 

Another separate issue was when I forgotten to specify the macro name in configuration and only had its namespace as in:
```
{:hook {:analyze-call {my-namespace hook-namespace/macro-name}}}
```

I can file a separate issue for this if it makes sense as it would really help if clj-kondo was flagging such errors (otherwise we need a tool to lint clj-kondo configuration :-)).

**Comment by borkdude** [comment]
That is expected since without a config, clj-kondo doesn't know at all that `def-whatever` is introducing bindings. `def-catch-all` is designed to not warn about anything but only capture that a var is introduced with a certain name.

**Comment by borkdude** [comment]
> I found that the same macro had two lint-as definitions - one inline and another in the global config.edn file where it was configured with clj-kondo.lint-as/def-catch-all.

Separate issue(s) welcome. Glad you figured it out.
